barman/cloud_providers/google_cloud_storage.py

```diff
@@ -333,3 +337,222 @@ def _delete_objects_batch(self, paths):
+                % (instance_name, zone, self.project)
+            )
+
+    def _get_disk_metadata(self, disk_name, zone):
+        try:
+            return self.disks_client.get(
+                disk=disk_name, zone=zone, project=self.project
+            )
+        except NotFound:
+            raise SnapshotBackupException(
+                "Cannot find disk with name %s in zone %s for project %s"
+                % (disk_name, zone, self.project)
+            )
+
+    def take_snapshot(self, backup_info, disk_zone, disk_name):
+        snapshot_name = "%s-%s-%s" % (
+            backup_info.server_name.lower(),
+            backup_info.backup_id.lower(),
+            disk_name,
+        )
+        _logger.info("Taking snapshot '%s' of disk '%s'", snapshot_name, disk_name)
+        resp = self.client.insert(
+            {
+                "project": self.project,
+                "snapshot_resource": {
+                    "name": snapshot_name,
+                    "source_disk": "projects/%s/zones/%s/disks/%s"
+                    % (
+                        self.project,
+                        disk_zone,
+                        disk_name,
+                    ),
+                },
+            }
+        )
+
+        _logger.info("Waiting for snapshot '%s' completion", snapshot_name)
+        resp.result()
+
+        if resp.done():
+            _logger.info("Snapshot '%s' completed", snapshot_name)
+            return snapshot_name
+        else:
+            raise CloudProviderError(
+                "Snapshot '%s' failed with error code %s: %s"
+                % (snapshot_name, resp.error_code, resp.error_message)
+            )
+
+    def take_snapshot_backup(self, backup_info, instance_name, zone, disks):
+        """Take a snapshot backup for the named instance."""
+        instance_metadata = self._get_instance_metadata(instance_name, zone)
+        snapshots = {}
+        for disk_name in disks:
+            disk_metadata = self._get_disk_metadata(disk_name, zone)
+            metadata = {
+                "block_size": disk_metadata.physical_block_size_bytes,
+                "size": disk_metadata.size_gb,
+            }
+            # Check disk is attached and find device name
+            attached_disks = [
+                d
+                for d in instance_metadata.disks
+                if d.source == disk_metadata.self_link
+            ]
+            if len(attached_disks) == 0:
+                raise SnapshotBackupException(
+                    "Disk %s not attached to instance %s" % (disk_name, instance_name)
+                )
+            elif len(attached_disks) > 1:
+                raise SnapshotBackupException(
+                    "Multiple disks matching name %s found attached to instance %s"
+                    % (disk_name, instance_name)
+                )
+            metadata["device"] = self.DEVICE_PREFIX + attached_disks[0].device_name
```

Comment:
What I had planned here was that the `device` field would always be a fully qualified path to the attached device, so consumers of the snapshots_info would not need to care about constructing such a path themselves - they can just use whatever is in the `device` field and not care how it was created.

However I agree with you that this presents a problem when a consumer wants to resolve this back to the `deviceName` as it known by GCP, but I'd still like to keep the fully qualified path in the metadata rather than requiring it to be composed at time-of-use.

What do you think about storing both the device_name, as retrieved from the GCP instance metadata, *and* the full device path?

Reply:
This looks like a great idea.